packages/blocks/src/api/serializer.js

```diff
@@ -220,7 +227,9 @@ export function getCommentAttributes( blockType, attributes ) {
 			// Ignore default value.
 			if (
 				'default' in attributeSchema &&
-				attributeSchema.default === value
+				( attributeSchema.default === value ||
+					( attributeSchema.type === 'object' &&
+						isEqual( attributeSchema.default, value ) ) )
```

Comment:
A similar change has been proposed in a PR at some point and was rejected (don't remember why, maybe performance). It would be good to try to find this PR and the related discussions :)

Reply:
Is this you had in mind: https://github.com/WordPress/gutenberg/pull/26162? It's a bit different but has to do with `object` default values.

Reply:
This one I think https://github.com/WordPress/gutenberg/pull/27348

Reply:
Isn't it a breaking change? It should be covered with a unit test if we decide to take this approach. What should we do about arrays?

Reply:
Of course I'll add tests if this will be the way to go. I think we could also handle `arrays` the same way but first only perform the simple equality value check to avoid the extra call for most cases (performance).

I'm not sure how it's a breaking change.

Reply:
Think about it this way, at the moment every object and array set as default is serialized in the HTML comment which means that when the default changes the block will use the value from the HTML comment. In case of simple types, if the default value changes the all blocks will get automatically updated to use the new default value.

Reply:
I'm not really sure if this would be a bad thing or breaking change though. Right now with the serialization in the HTML comment of `objects` and/or `arrays` the user sees the result and applies changes if wants. If no changes applied, it means that they are okay with the provided values, so even if the default changes they will not have any problem.

I'm not 100% clear on how to best handle this, but if my above assumption is correct it might be more confusing what happens now with simple type attributes.

Reply:
I'm not entirely sure how this change relates to this PR and whether it's mandatory but I personally don't have big objections here if it's mandatory, it seems the discussions on the linked PR were mostly about performance impact.

Reply:
> I'm not entirely sure how this change relates to this PR and whether it's mandatory but I personally don't have big objections here if it's mandatory, it seems the discussions on the linked PR were mostly about performance impact.

Performance was one concern, but the other was the semantic distinction between no value and empty value (i.e. `undefined !== {}`).

But yeah, in this case I'm not sure how it relates to the current PR — what am I missing?

Reply:
It all started from the deprecation snapshot here: https://github.com/WordPress/gutenberg/pull/30804#discussion_r613084379.

If we are okay to leave the snapshot as it was with the serialized `query` object, I can remove this change and move forward this PR with the tests fixing/check.

Reply:
Thanks for the link, I see it now. I don't think this is an easy question. We'd have to weigh the trade-offs: 
* Using `isEqual` in the serialiser would simplify the case of attributes with deep-valued defaults, which would in a sense "fix" how default values.
* On the other hand, the serialiser has worked like this for four years now (and two in Core), and the fix might cause unexpected changes in third-party blocks who have come to rely on this quirk.
* I feel less strongly today about what I said in December in https://github.com/WordPress/gutenberg/pull/27348#discussion_r543608248, but there is still a point to consider there.
* Finally, obviously, there's the question of performance associated with deep comparison. I don't expect this to be a major factor, in practice, but we would have to test.

I'm in favour of keeping the focus of this PR on the alignment win, and leaving the serialisation matter aside.

Reply:
> On the other hand, the serialiser has worked like this for four years now (and two in Core), and the fix might cause unexpected changes in third-party blocks who have come to rely on this quirk.

That is my main concern with this proposal. Another thing I worry about is that when comparing arrays it's unclear if the order of items should be taken into account. I guess, it's better to leave it unchanged and just open a follow-up issue to document better the current behavior in the documentation.

Reply:
>On the other hand, the serialiser has worked like this for four years now (and two in Core), and the fix might cause unexpected changes in third-party blocks who have come to rely on this quirk.

>I guess, it's better to leave it unchanged and just open a follow-up issue to document better the current behavior in the documentation.

I agree. This is a totally different and bigger issue that needs resolving on its own. I'll revert the related change.